### PR TITLE
Add goto issue id function

### PR DESCRIPTION
**Pull request by tyroneyeh**
for https://github.com/go-gitea/gitea/issues/4109#issuecomment-1527104992 

Supports format:
`#1234`
`Org/Repo#1234`

### CODE REVIEW

**Comment by wxiaoguang** [comment]
I definitely like this idea more than adding the issue index into search engine (#19646)

Some more background: some people suggested to support the number without `#` prefix (the reasons is that many users do not have enough technical background).

And I guess we do not need to change the "search" behavior. So I think we can show a "goto" button dynamically like this, then all users could be satisfied.

```

[ input: #123 ]  [button:goto] [button:search] 

[ input: 1234 ]  [button:goto] [button:search] 

[ input:  Org/Repo#1234 ]  [button:goto] [button:search] 

[ input: other input ]  [button:search] 

```

**Comment by tyroneyeh** [comment]
Added dynamic hash-tag button for go to function

image
image

**Comment by wxiaoguang** [comment]
Some quick reviews:

1. There is `parseIssueHref`, it's covered by test, maybe we do not need to use new regexp to process the issue index
2. According to https://docs.gitea.io/en-us/contributing/guidelines-frontend/
    * HTML ids and classes should use kebab-case
    * Show/hide: https://docs.gitea.io/en-us/contributing/guidelines-frontend/#showhide-elements
3. It seems not good to have two `type="submit"` buttons in one form, especially the first one won't submit the form
4. The `aria-label=` is not accurate for the new button, it's not for "search", but for "goto"
5. The new "goto" button could have a tooltip, otherwise some users may be confused about it's purpose (with a correct tooltip, then no need to add `aria-label`)

**Comment by silverwind** [comment]
Please use camelCase for variable names and don't be afraid to use longer names. I don't know what `pathslahnum` is supposed to mean.

File: templates/repo/issue/search.tmpl
Comment (silverwind) [inline templates/repo/issue/search.tmpl]: Use `gt-hidden` class to hide and then use `showElem`/`hideElemn`/`toggleElem` helper function from `utils/dom.js` to change visibility.

**Comment by tyroneyeh** [comment]
> Some quick reviews:
> 
> 1. There is `parseIssueHref`, it's covered by test, maybe we do not need to use new regexp to process the issue index
Please help modify it, thanks

> 2. According to https://docs.gitea.io/en-us/contributing/guidelines-frontend/
>    
>    * HTML ids and classes should use kebab-case
>    * Show/hide: https://docs.gitea.io/en-us/contributing/guidelines-frontend/#showhide-elements
Modified

> 3. It seems not good to have two `type="submit"` buttons in one form, especially the first one won't submit the form
Reuse submitted events

> 4. The `aria-label=` is not accurate for the new button, it's not for "search", but for "goto"
Modified

> 5. The new "goto" button could have a tooltip, otherwise some users may be confused about it's purpose (with a correct tooltip, then no need to add `aria-label`)
The goto button is hidden by default, and generally you don’t know that there is this function. It may be changed to the preset disabled, and it will be enabled after the input conforms to the format. It may be more intuitive.

**Comment by wxiaoguang** [comment]
> > Some quick reviews:
> > 1. There is `parseIssueHref`, it's covered by test, maybe we do not need to use new regexp to process the issue index
> 
> Please help modify it, thanks
> 

Hmm ... I don't have writer permission at the moment (I quited manager team some time ago, just re-joint recently, so haven't got writer permission)

But `parseIssueHref` is covered by tests, the tests show how it works (if it helps)


> > 3. It seems not good to have two `type="submit"` buttons in one form, especially the first one won't submit the form
> 
> Reuse submitted events
> 

I don't see why it is necessary. Removing the type=submit from goto button, I guess the form submit event still works?


> > 5. The new "goto" button could have a tooltip, otherwise some users may be confused about it's purpose (with a correct tooltip, then no need to add `aria-label`)
> 
> The goto button is hidden by default, and generally you don’t know that there is this function. It may be changed to the preset disabled, and it will be enabled after the input conforms to the format. It may be more intuitive.

"hidden by default" I think it is fine, adding a tooltip is good enough

File: templates/repo/issue/search.tmpl
Comment (wxiaoguang) [inline templates/repo/issue/search.tmpl]: Using `data-tooltip-content="{{.locale.Tr "explore.goto"}}"` is enough (no need to write aria-label)

File: web_src/js/features/repo-issue.js
Comment (wxiaoguang) [inline web_src/js/features/repo-issue.js]: use `input` event , it works for mouse-pasting

**Comment by tyroneyeh** [comment]
> I don't see why it is necessary. Removing the type=submit from goto button, I guess the form submit event still works?

I listen to the event processing sent by the form. If the form is not submit, there will be no action. If you enter #1234 in the input field and press enter, the submit event will also be trigger

**Comment by wxiaoguang** [comment]
> > I don't see why it is necessary. Removing the type=submit from goto button, I guess the form submit event still works?
> 
> I listen to the event processing sent by the form. If the form is not submit, there will be no action. If you enter #1234 in the input field and press enter, the submit event will also be trigger

I see, you are right, I understand what the code does now. ps: all these `type=submit`  could be removed (but not a must), according to MDN: the buttons have `type=submit` by default 🤣  https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button

**Comment by tyroneyeh** [comment]
> I see, you are right, I understand what the code does now. ps: all these `type=submit` could be removed (but not a must), according to MDN: the buttons have `type=submit` by default  https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button

ok, I removed it.

File: options/locale/locale_en-US.ini
Comment (silverwind) [inline options/locale/locale_en-US.ini]: Would say make this `go_to = Go to`. "Goto" always makes me think of the same-named statement in some programming languages, so clear english seems better.

**Comment by silverwind** [comment]
Counting slashes in the path seems to brittle. There is a rarely used method of installation where users install Gitea on a subdirectory, e.g. `https://example.com/gitea` and I assume this slash counting would break in such a scenario.

If possible re-use `parseIssueHref`  or add your own parser function in `utils.js`, with tests for it in `utils.test.js`.

**Comment by tyroneyeh** [comment]
> Counting slashes in the path seems to brittle. There is a rarely used method of installation where users install Gitea on a subdirectory, e.g. `https://example.com/gitea` and I assume this slash counting would break in such a scenario.
> 
> If possible re-use `parseIssueHref` or add your own parser function in `utils.js`, with tests for it in `utils.test.js`.

I improved, please help me to check again
Thanks

File: web_src/js/features/repo-issue.js
Comment (lunny) [inline web_src/js/features/repo-issue.js]: use `appSubUrl` directly.

File: web_src/js/features/repo-issue.js
Comment (lunny) [inline web_src/js/features/repo-issue.js]: I think it has been imported in line 9.

File: web_src/js/features/repo-issue.js
Comment (lunny) [inline web_src/js/features/repo-issue.js:644]: class `list-header-search` has also been used on milestone search but only issues search are expected.

Comment (tyroneyeh) [reply web_src/js/features/repo-issue.js:644]: This on milestones no effect

**Review by wxiaoguang — APPROVED** [review]
Some details do not look good to me (including: how the query text is parsed, when the button should be shown).

But I can't help to edit this PR. So, put my approval here.

After merging, I will propose new PRs for do more improvements.

**Review by lunny — APPROVED** [review]
The idea is good and the code could be improved by future PRs.

**Comment by delvh** [comment]
I've just tested this functionality on try.gitea.io, and I have some reservations regarding it's current implementation:
1. It's not a good UI design to hide functionality like this completely. Typically, you should simply disable a button when it cannot be used (yet)
2. I am not a fan of detecting whether we are on the global issues page or not by checking if an element is present. This could have been a parameter, or some other form of detection that doesn't break as likely, i.e. using `window.config.pageData`.
3. The full `owner/repo#nr` syntax seems buggy: When I enter `delvh/test#3` on https://try.gitea.io/delvh/test/issues, for example, I get redirected to https://try.gitea.io/delvh/test/delvh/test/issues/3.
4. When I click on the `#` button and immediately return to the search page, the text will still be shown in the textarea **but** the button will only be shown when I manually change the text. We should also check if the initial text matches the pattern.

**Comment by wxiaoguang** [comment]
> Some details do not look good to me (including: how the query text is parsed, when the button should be shown).
> But I can't help to edit this PR. So, put my approval here.
> After merging, I will propose new PRs for do more improvements.

Yup, it needs improvements

**Comment by wxiaoguang** [comment]
->  Improve "goto issue by number" button #24577

**Pull request merged**
